src/main/java/com/google/research/bleth/simulator/SimulationMetadata.java

```diff
@@ -108,17 +110,27 @@ public static SimulationMetadata read(String simulationId) {
 
     /**
      * Read all existing SimulationMetadata entites from the db, and return them as a hashmap.
+     * @param sortProperty is the name of the property to sort the results by (optional).
+     * @param sortDirection is the direction of the sort (optional).
      * @return a hashmap mapping a simulationId to the corresponding SimulationMetadata object.
      */
-    public static ImmutableMap<String, SimulationMetadata> listSimulations() {
-        ImmutableMap.Builder<String, SimulationMetadata> simulations = new ImmutableMap.Builder<>();
+    public static LinkedHashMap<String, SimulationMetadata> listSimulations(Optional<String> sortProperty,
```

Comment:
Why is this a `Map` at all? Does any caller ever invoke `.get(foo)` on the returned value? (The only caller I see is the servlet, and it only iterates over the results)

Prefer to just change this method to return a `List<SimulationMetadata>`, which is precisely what the caller actually uses this for.

Reply:
The reason is that the id is not a property of the `SimulationMetadata` class; the id is the return value of the `SimulationMetadata.write()` method. In order to apply the change you have suggested, we need to add an `id` property to the `SimulationMetadata` class, and update it inside the `SimulationMetadata.write()`.

This change sound relatively simple, but I wonder if the id _should_ actually be a property of the `SimulationMetadata` class - what should happen if the same `SimulationMetadata` object calls his `write()` method multiple time? This might happen if we use the same builder to construct multiple simulation with the same setup, for example.

Reply:
Plus: this key-value struct is being used in client-side code, where this map is being parsed as a JSON object (a list can also be parsed as JSON, but it is easier to deal with when the key is the simulation id and the value is another JSON object storing all other simulation's metadata parameters).

Reply:
Still - at least here, you're not truly using this as a "map" per se; it's more of a "list of key-value pairs".

My usual recommendation in this case would be to create yet another class that holds a String key and a `SimulationMetadata` element, and return a list of those.

Reply:
No problem at all, but may I ask why? Isn't this exactly what `LinkedHashMap` does?

Reply:
`LinkedHashMap` is still a `Map`, and the assumption of the reader is that this is important (i.e. that callers need to use the `Map`-specific APIs, such as `get(key)` and `put(key, value)`. Here you're just using this class for its convenience value of having an `entrySet()` - which isn't a core part of what a `Map` is.

Reply:
OK; I have to eat a hat.

I just looked at the internal guide to "choosing simpler types" - and it does recommend to use a `Map` instead of an `Iterable<Pair<Foo, Bar>>` or equivalent.

One thing it does mention is that `ImmutableMap` *also* provides deterministic ordering (i.e., its `entrySet` is ordered in the same order of elements added to the `builder()`). So - just use `ImmutableMap`, and be done with it.

¯\_(ツ)_/¯